src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs

```diff
@@ -0,0 +1,334 @@
+            in T? rootValue,
+            object? rootValueBoxed = null,
+            bool isInvokedByPolymorphicConverter = false)
+        {
+            Debug.Assert(IsConfigured);
+            Debug.Assert(rootValueBoxed is null || rootValueBoxed is T);
+
+            if (CanUseSerializeHandler)
+            {
+                // Short-circuit calls into SerializeHandler, if supported.
+                // Even though this is already handled by JsonMetadataServicesConverter,
+                // this avoids creating a WriteStack and calling into the converter infrastructure.
+
+                Debug.Assert(SerializeHandler != null);
+                Debug.Assert(Options.SerializerContext?.CanUseSerializationLogic == true);
+                Debug.Assert(Converter is JsonMetadataServicesConverter<T>);
+
+                SerializeHandler(writer, rootValue!);
+                writer.Flush();
+            }
+            else if (
+#if NETCOREAPP
+                !typeof(T).IsValueType &&
+#endif
+                Converter.CanBePolymorphic &&
+                rootValue is not null &&
+                Options.TryGetPolymorphicTypeInfoForRootType(rootValue, out JsonTypeInfo? derivedTypeInfo))
+            {
+                Debug.Assert(typeof(T) == typeof(object));
+                derivedTypeInfo.SerializeAsObject(writer, rootValue, isInvokedByPolymorphicConverter: true);
+                // NB flushing is handled by the derived type's serialization method.
+            }
+            else
+            {
+                WriteStack state = default;
+                state.Initialize(this, rootValueBoxed, isPolymorphicRootValue: isInvokedByPolymorphicConverter);
+
+                bool success = EffectiveConverter.WriteCore(writer, rootValue!, Options, ref state);
+                Debug.Assert(success);
+                writer.Flush();
+            }
+        }
+
+        // Root serialization method for async streaming serialization.
+        internal async Task SerializeAsync(
+            Stream utf8Json,
+            T? rootValue,
+            CancellationToken cancellationToken,
+            object? rootValueBoxed = null,
+            bool isInvokedByPolymorphicConverter = false)
+        {
+            Debug.Assert(IsConfigured);
+            Debug.Assert(rootValueBoxed is null || rootValueBoxed is T);
+
+            if (CanUseSerializeHandlerInStreaming)
+            {
+                // Short-circuit calls into SerializeHandler, if the `CanUseSerializeHandlerInStreaming` heuristic allows it.
+
+                Debug.Assert(SerializeHandler != null);
+                Debug.Assert(Options.SerializerContext?.CanUseSerializationLogic == true);
+                Debug.Assert(Converter is JsonMetadataServicesConverter<T>);
+
+                using var bufferWriter = new PooledByteBufferWriter(Options.DefaultBufferSize);
+                using var writer = new Utf8JsonWriter(bufferWriter, Options.GetWriterOptions());
```

Comment:
Can this use `Utf8JsonWriterCache.RentWriterAndBuffer` instead of allocating a new `Utf8JsonWriter` every time?

Allocating a new Utf8JsonWriter seems to be one of the bigger allocs in ASP.NET - https://github.com/dotnet/aspnetcore/issues/45332.

In https://github.com/aspnet/Benchmarks/pull/1772/ we are manually caching the `Utf8JsonWriter` in the benchmark.

cc @DamianEdwards @BrennanConroy

Reply:
We can do that because we use the sync APIs right? Once you go async it'll mess up the current Utf8JsonWriterCache implementation I think.

Reply:
Can we just reuse the `Utf8JsonWriter` and not the `PooledByteBufferWriter`? The `Utf8JsonWriter` doesn't seem to be used asynchronously here.

Reply:
Yep, that can be done. I've updated the branch to include that change.